sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -9,91 +9,188 @@
+        &manifest::ROM_EXT_MANIFEST_IDENTIFIER,
+        parse_hex_str("0x4552544f")?,
+    )?;
+
+    image.set_manifest_field(
+        &manifest::ROM_EXT_IMAGE_VERSION,
+        parse_hex_str("0xa8ad8002")?,
+    )?;
+
+    // TODO: Do we need these fields?
+    let extensions = [
+        (
+            &manifest::ROM_EXT_EXTENSION0_OFFSET,
+            &manifest::ROM_EXT_EXTENSION0_CHECKSUM,
+        ),
+        (
+            &manifest::ROM_EXT_EXTENSION1_OFFSET,
+            &manifest::ROM_EXT_EXTENSION1_CHECKSUM,
+        ),
+        (
+            &manifest::ROM_EXT_EXTENSION2_OFFSET,
+            &manifest::ROM_EXT_EXTENSION2_CHECKSUM,
+        ),
+        (
+            &manifest::ROM_EXT_EXTENSION3_OFFSET,
+            &manifest::ROM_EXT_EXTENSION3_CHECKSUM,
+        ),
+    ];
+    for fields in &extensions {
+        image.set_manifest_field(fields.0, parse_hex_str("0xdeadbeef")?)?;
+        image.set_manifest_field(fields.1, parse_hex_str("0x8badbeef")?)?;
+    }
+    image.set_manifest_field(
+        &manifest::ROM_EXT_USAGE_CONSTRAINTS,
+        std::array::IntoIter::new(
+            [0xa6; manifest::ROM_EXT_USAGE_CONSTRAINTS.size_bytes],
+        ),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_PERIPHERAL_LOCKDOWN_INFO,
+        std::array::IntoIter::new(
+            [0xa5; manifest::ROM_EXT_PERIPHERAL_LOCKDOWN_INFO.size_bytes],
+        ),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_IMAGE_TIMESTAMP,
+        parse_hex_str("0x7766554433221100")?,
+    )?;
+
+    Ok(())
+}
 
-    // The whole image has been updated and signed, write the result to disk.
-    image.write_file();
+/// Calculates the signature for the signed portion of an image.
+fn calculate_image_signature(
+    image: &Image,
+    key: impl AsRef<Path>,
+) -> Result<ImageSignature> {
+    let key = fs::read(&key).with_context(|| {
+        format!("Failed to read key from `{}`.", key.as_ref().display())
+    })?;
+    let key =
+        PrivateKey::parse_from_der(&key).context("Failed to parse the key.")?;
+    let sig = ImageSignature::sign(&key, image.signed_bytes())
+        .context("Failed to sign the image.")?;
+    Ok(sig)
 }
 
-/// Generates the device usage value.
-///
-/// This value is extrapolated from the ROM_EXT manifest usage_constraints
-/// field, and does not reside in the ROM_EXT manifest directly.
-pub fn device_usage_value(path: &Path) -> Vec<u8> {
-    let _usage_constraints =
-        fs::read(path).expect("Failed to read usage constraints!");
-
-    // TODO: generate the device_usage_value from usage_constraints.
-    //       meanwhile use a "dummy" hard-coded vector.
-    vec![0xA5; 1024]
+/// Updates the signature of an image.
+fn update_image_signature(
+    image: &mut Image,
+    signature: ImageSignature,
+) -> Result<()> {
+    image.set_manifest_field(
+        &manifest::ROM_EXT_IMAGE_SIGNATURE,
+        signature.bytes().iter().rev().cloned(),
+    )?;
```

Comment:
I think you can just return this directly?

Reply:
The return type is `anyhow::Result<()>`.  In my own experience, if the function being called returns a different type of result (e.g. `io::Result<T>`), then the compiler will complain about a type mismatch.

However, if one uses `?`, the compiler will figure out if it can perform an `into()` to convert the result to the desired type.  Debatable as to whether one should use `?` or call `into()` directly.

Reply:
I also get a `mismatched types` if I try to return this directly. IIUC, `?` handles this conversion but I have to wrap the return value in `Ok()`.